### ISSUE TITLE
[Proposal] Add WaitUntil block

### ISSUE THREAD

**Issue by fflorent** [junegunn/vader.vim#69]
I would love to be able to support asynchronicity through a `WaitUntil [autocmd-events] [(comments)]` block, like this:

````
Execute (some asynchronous job):
  MyAsynchronousCommand

WaitUntil BufNew,BufRead (new window pops up):
  bufname('.') ==# '/tmp/foo' # Wait until a new buffer being created is /tmp/foo
  # Other conditions here would be treated as AND conditions

Expect:
  some content in /tmp/foo
````

If the `autocmd-events` are omitted, the WaitUntil would listen by default to a set of events (to determine) like BufNew, BufRead, BufWrite, CursorMove, ...

This would especially be helpful for asynchronous NeoVim plugins.

Thoughts?

(If you think that is a good idea but lack time, I can try to propose you a patch)

Florent

**Comment by junegunn** [comment]
The test runner of Vader is inherently synchronous and making it asynchronous is not going to trivial. Basically we'll have to restructure the test runner so that it can be paused and resumed during its execution (Correct me if I'm wrong).

I haven't really thought about how we should write tests for asynchronous, interactive features. `WaitUntil` construct you suggested can be one way to do it, but it looks like it does not cover all possible cases.
- The feature may not involve any autocmd. (Hmm, but then how we're going to check its progress?)
- A test case may require multiple asynchronous interactions, which means multiple waits in a Do or Execute block
    - e.g. Start fzf, wait for some text to appear, type in a query string, wait for the filtered result, move cursor, type some key, then check the result

**Comment by fflorent** [comment]
> The test runner of Vader is inherently synchronous and making it asynchronous is not going to trivial. Basically we'll have to restructure the test runner so that it can be paused and resumed during its execution (Correct me if I'm wrong).

That's very true. Currently I am playing with the code in order to write a system of callbacks. 

Unfortunately, that's not very nice so far. I'd like to continue that work so we can see how it evolves.

> The feature may not involve any autocmd. (Hmm, but then how we're going to check its progress?)

In that particular cases (e.g. wait until a background HTTP request is done), a developer may raise a custom event in its code, then do:

````
WaitUntil User MyCustomEvent:
  ...
````

> * A test case may require multiple asynchronous interactions, which means multiple waits in a Do or Execute block
> 
>   * e.g. Start fzf, wait for some text to appear, type in a query string, wait for the filtered result, move cursor, type some key, then check the result

Well, that are 4 `WaitUntil`s in that case. If that's really a drawback, I see these solutions:
* split the test files into small ones (unitary)
* or make util test files that can be included

For the last option:
````
Execute:
  let g:test#text_to_type = "some text"
Include: util/type_text.vader
````

However, I see a new drawback now: how can we loop with asynchronicity? E.g. one wants to trigger moves / actions until a condition is met (i.e. series of `Do` / `Execute` + `WaitUntil`).

Anyway, may I propose a progress with the idea as a first step, realize some improvements later before you decide to merge or not? We may have clearer ideas of the limitations / opportunities while implementing the solution.

Also don't hesitate to suggest a better idea, I would be very happy to implement it :).

Florent

**Comment by junegunn** [comment]
Feel free to explore. It's an open source project after all. You can still use your fork for testing your projects even if it's not merged.

I'm interested in testing Neovim plugin of fzf. On plain vim, fzf runs on a new tmux pane via synchronous `system()` call, so there's nothing we can do. But on the other hand on Neovim, fzf runs on a terminal buffer asynchronously, so there's a chance we can test it somehow. However fzf is just an external program so it does not fire any autocmd. Maybe we could consider using `CursorHold` (with short `updatetime`) to periodically monitor its progress.

As noted above, a typical test case for fzf logically requires multiple waits and we cannot break it into a series of separate test cases, i.e. multiple `Execute`s or `Do`s. Also note that Vader resets the workbench buffer between every test case. Basically I need to be able to write something like the following:

```
Execute (Test Files command):
   Files /tmp
   wait until the last line of the terminal buffer contains "> " (with some timeout)
   call feedkeys('query')
   wait until the second to the last line the terminal buffer contains "1/"
   call feedkeys("\<c-t>")
   AssertEqual ...

Execute (Test Buffers command):
  ...
```

And actually that's how I've written test cases for fzf itself, using tmux:
https://github.com/junegunn/fzf/blob/master/test/test_go.rb#L404-L413

**Comment by fflorent** [comment]
Above are some POC I made lately... The last commit succeeds in listening to events and resuming the execution, also I still have to fix some issues.

I also wonder if there are simpler solutions for the callbacks (the first commit). I am not very used to VimScript, as you may notice :).

Florent

**Comment by neersighted** [comment]
As I have recently encountered issues testing async plugins, this is high on my list.
Any progress on this, @fflorent? I'd love to help if needed.

**Comment by fflorent** [comment]
Not really. I don't have any idea how to implement the syntax described by @junegunn above. Feel free to take the issue :).

Florent

**Comment by junegunn** [comment]
I'll be looking into this as I need it for https://github.com/junegunn/vim-plug/pull/534

**Comment by blueyed** [comment]
btw: for Neomake I am using custom functions/commands with "sleep": https://github.com/neomake/neomake/blob/7502e43d8e60f3dfe2521e1b0df5c4f712656949/tests/_setup.vader#L4

**Comment by junegunn** [comment]
@blueyed Ah, that should work. For some reason I was fixated on the idea the whole time that polling with sleep wouldn't work, but now I realized Vimscript does relinquish control on sleep. Bummer.

**Comment by blueyed** [comment]
> Vimscript does relinquish control on sleep

That requires a certain Vim 8 patch level though IIRC.

**Comment by junegunn** [comment]
@blueyed Ah didn't know that, thanks. By the way, are you closely following the development of Vim? Any idea what might have caused this? https://github.com/junegunn/vim-plug/issues/537
